src/common/cockpitframe.c

```diff
@@ -0,0 +1,239 @@
+
+  while (length > 0)
+    {
+      res = write (fd, data, length);
+      if (res < 0)
+        {
+          if (errno != EAGAIN && errno != EINTR)
+            return -1;
+        }
+      else
+        {
+          data += res;
+          length -= res;
+          written += res;
+        }
+    }
+
+  return written;
+}
+
+ssize_t
+cockpit_frame_write (int fd,
+                     unsigned char *input,
+                     size_t length)
+{
+  char *prefix = NULL;
+  ssize_t ret = -1;
+  int errn = 0;
+
+  assert (length > 0);
+  assert (input != NULL);
+
+  if (asprintf (&prefix, "%u\n", (unsigned int)length) < 0)
+    {
+      errn = ENOMEM;
+      goto out;
+    }
+
+  ret = write_all (fd, (unsigned char *)prefix, strlen (prefix));
+  if (ret > 0)
+    ret = write_all (fd, input, length);
+  if (ret < 0)
+    errn = errno;
+
+out:
+  free (prefix);
+  if (ret < 0)
+    errno = errn;
+  return ret;
+}
+
+static void *
+xrealloc (void *old,
+          size_t length)
+{
+  void *data = realloc (old, length);
+  if (!data && length > 0)
+    free (old);
+  return data;
+}
+
+ssize_t
+cockpit_frame_read (int fd,
+                    unsigned char **output)
+{
+  ssize_t size = 0;
+  size_t skip;
+  ssize_t res;
+  int errn = 0;
+  ssize_t ret = -1;
+
+  unsigned char *buf = NULL;
+  size_t buflen = 0;
+  size_t allocated = 0;
+
+  while (size == 0 || buflen < size)
+    {
+      /* Reallocate */
+      if (buflen + 1 > allocated)
+        {
+          allocated = size;
+          if (allocated < buflen + 128)
+            allocated = buflen + 128;
+          buf = xrealloc (buf, allocated);
+          if (!buf)
+            {
+              errn = ENOMEM;
+              goto out;
+            }
+        }
+
+      res = read (fd, buf + buflen, 1);
+      if (res < 0 && errno == ECONNRESET && buflen == 0)
+        res = 0;
+      if (res < 0)
+        {
+          /* A read failure */
+          if (errno != EINTR || errno != EAGAIN)
+            {
+              errn = errno;
+              goto out;
+            }
+        }
+      else if (res == 0)
+        {
+          free (buf);
+
+          /* No message parsed, but also no data received */
+          if (size == 0 && buflen == 0)
+            ret = 0;
+          else
+            errn = EBADMSG;
+          goto out;
+        }
+      else if (res > 0)
+        {
+          buflen += 1;
+        }
+
+      /* Parse the length if necessary */
+      if (size == 0)
+        {
+          assert (buf != NULL);
+          size = cockpit_frame_parse (buf, buflen, &skip);
+          if (size > 0)
+            {
+              assert (buflen >= skip);
+              if (buflen > skip)
+                memmove (buf, buf + skip, buflen - skip);
+              buflen -= skip;
+            }
+        }
+
+      if (size < 0)
+        {
+          errn = EBADMSG;
+          goto out;
+        }
+    }
+
+  if (output)
+    {
+      *output = buf;
+      buf = NULL;
+    }
+  ret = size;
```

Comment:
This looks wrong. In the successful case (above `if`) you then run into `out:` which frees `buf` and thus the returned `*output`. I suggest to `return size` in the above `if` and renaming `out:` to `err:` to clarify it's only the error path.